trl/commands/cli_utils.py

```diff
@@ -152,88 +163,126 @@ class ChatArguments:
 
 class TrlParser(HfArgumentParser):
     """
-    The TRL parser parses a list of parsers (TrainingArguments, trl.ModelConfig, etc.), creates a config
-    parsers for users that pass a valid `config` field and merge the values that are set in the config
-    with the processed parsers.
+    A subclass of [`transformers.HfArgumentParser`] designed for parsing command-line arguments with dataclass-backed
+    configurations, while also supporting configuration file loading and environment variable management.
 
     Args:
-        parsers (`List[argparse.ArgumentParser]`):
-            List of parsers.
-        ignore_extra_args (`bool`):
-            Whether to ignore extra arguments passed by the config
-            and not raise errors.
+        dataclass_types (`Union[DataClassType, Iterable[DataClassType]]`):
+            The dataclass types to use for argument parsing.
+        **kwargs:
+            Additional keyword arguments passed to the [`transformers.HfArgumentParser`] constructor.
+
+    Examples:
+
+    ```yaml
+    # config.yaml
+    env:
+        VAR1: value1
+    arg1: 23
+    ```
+
+    ```python
+    # main.py
+    import os
+    from dataclasses import dataclass
+    from trl import TrlParser
+
+    @dataclass
+    class MyArguments:
+        arg1: int
+        arg2: str = "alpha"
+
+    parser = TrlParser(dataclass_types=[MyArguments])
+    training_args = parser.parse_args_and_config()
+
+    print(training_args, os.environ.get("VAR1"))
+    ```
+
+    ```bash
+    $ python main.py --config config.yaml
+    (MyArguments(arg1=23, arg2='alpha'),) value1
+
+    $ python main.py --arg1 5 --arg2 beta
+    (MyArguments(arg1=5, arg2='beta'),) None
+    ```
     """
 
-    def __init__(self, parsers, ignore_extra_args=False):
-        super().__init__(parsers)
-        self.yaml_parser = YamlConfigParser()
-        self.ignore_extra_args = ignore_extra_args
+    def __init__(
+        self,
+        dataclass_types: Union[DataClassType, Iterable[DataClassType]],
+        ignore_extra_args: Optional[bool] = None,
+        **kwargs,
+    ):
+        super().__init__(dataclass_types=dataclass_types, **kwargs)
+        if ignore_extra_args is not None:
+            warnings.warn(
+                "The `ignore_extra_args` parameter is deprecated and will be removed in version 0.14. "
```

Comment:
Is this deprecated to enforce strict typing between a dataclass and the runtime config? I'm mostly wondering if this was included by design in the original parser for some specific use cases. In any case, let's deprecate and see if we get reports between now and the deprecation

Reply:
Good point! done in d378afa

Reply:
Initially I deprecated it because I didn't understand what it was doing. It seemed to be no-op. But now I understand: When you have an argument in `config.yaml` that is not passed (belongs to no dataclass, then it returns a value error. Unless you set this arg to True:

```yaml
# config.yml
output_dir: "output_dir"
extra_arg: "This is not a recognized argument"
```

```python
from trl import TrlParser
from transformers import TrainingArguments

parser = TrlParser((TrainingArguments,), ignore_extra_args=False)
args = parser.parse_args_and_config()  # raise a ValueError

parser = TrlParser((TrainingArguments,), ignore_extra_args=True)
args = parser.parse_args_and_config()
print(args)
# (TrainingArguments(_n_gpu=1, ...), Namespace(extra_arg='This is not a recognized argument'))
```

Finally, it's mostly the same as `return_remaining_strings` in `parse_args_and_config` but for the config file. What I'll do:

Keep this arg and make it functional; deprecate it in favour of `return_remaining_strings`. Something like

```python
from trl import TrlParser
from transformers import TrainingArguments

parser = TrlParser((TrainingArguments,))
args = parser.parse_args_and_config()  # raise a ValueError

parser = TrlParser((TrainingArguments,), ignore_extra_args=True)
args = parser.parse_args_and_config()
print(args)
# Warning, this is depracated, use `return_remaining_strings`` instead
# (TrainingArguments(_n_gpu=1, ...), Namespace(extra_arg='This is not a recognized argument'))

parser = TrlParser((TrainingArguments,))
args = parser.parse_args_and_config(return_remaining_strings=True) 
print(args)
# (TrainingArguments(_n_gpu=1, ...), Namespace(extra_arg='This is not a recognized argument'))
```

Reply:
Done in 4f87b75